src/main/java/com/epam/aidial/core/service/ResourceService.java

```diff
@@ -0,0 +1,401 @@
+
+            if (meta.getType() != StorageType.BLOB) {
+                return new ResourceFolderMetadata(description);
+            }
+
+            Long createdAt = null;
+            Long updatedAt = null;
+
+            if (metadata != null) {
+                createdAt = metadata.containsKey("created_at") ? Long.parseLong(metadata.get("created_at")) : null;
+                updatedAt = metadata.containsKey("updated_at") ? Long.parseLong(metadata.get("updated_at")) : null;
+            }
+
+            if (createdAt == null && meta.getCreationDate() != null) {
+                createdAt = meta.getCreationDate().getTime();
+            }
+
+            if (updatedAt == null && meta.getLastModified() != null) {
+                updatedAt = meta.getLastModified().getTime();
+            }
+
+            return new ResourceItemMetadata(description).setCreatedAt(createdAt).setUpdatedAt(updatedAt);
+        }).toList();
+
+        return new ResourceFolderMetadata(descriptor, resources).setNextToken(set.getNextMarker());
+    }
+
+    private ResourceItemMetadata getItemMetadata(ResourceDescription descriptor) {
+        String redisKey = redisKey(descriptor);
+        String blobKey = blobKey(descriptor);
+        Result result = redisGet(redisKey, false);
+
+        if (result == null) {
+            result = blobGet(blobKey, false);
+        }
+
+        if (!result.exists) {
+            return null;
+        }
+
+        return new ResourceItemMetadata(descriptor)
+                .setCreatedAt(result.createdAt)
+                .setUpdatedAt(result.updatedAt);
+    }
+
+    @Nullable
+    public String getResource(ResourceDescription descriptor) {
+        String redisKey = redisKey(descriptor);
+        Result result = redisGet(redisKey, true);
+
+        if (result == null) {
+            try (var lock = lockService.lock(redisKey)) {
+                result = redisGet(redisKey, true);
+
+                if (result == null) {
+                    String blobKey = blobKey(descriptor);
+                    result = blobGet(blobKey, true);
+                    redisPut(redisKey, result);
+                }
+            }
+        }
+
+        return result.exists ? result.body : null;
+    }
+
+    public boolean putResource(ResourceDescription descriptor, String body) {
+        String redisKey = redisKey(descriptor);
+        String blobKey = blobKey(descriptor);
+
+        try (var lock = lockService.lock(redisKey)) {
+            Result result = redisGet(redisKey, false);
+            if (result == null) {
+                result = blobGet(blobKey, false);
+            }
+
+            long updatedAt = time();
+            long createdAt = result.exists ? result.createdAt : updatedAt;
+            redisPut(redisKey, new Result(body, createdAt, updatedAt, false, true));
+
+            if (!result.exists) {
+                blobPut(blobKey, "", createdAt, updatedAt);
```

Comment:
Can you please add one line comment that we need it because we don't cache listings.

Reply:
added